spicy/runtime/src/parser.cc

```diff
@@ -194,19 +208,24 @@ hilti::rt::Bytes detail::extractBytes(hilti::rt::ValueReference<hilti::rt::Strea
                                       const hilti::rt::StrongReference<spicy::rt::filter::detail::Filters>& filters) {
     if ( eod_ok )
         detail::waitForInputOrEod(data, cur, size, filters);
-    else
-        detail::waitForInput(data, cur, size, hilti::rt::fmt("expected %" PRIu64 " bytes", size), location, filters);
+    else if ( ! detail::waitForInputNoThrow(data, cur, size, filters) ) {
+        auto msg = hilti::rt::fmt("expected %" PRIu64 " bytes (%" PRIu64 " available)", size, cur.size());
+        throw ParseError(msg, std::string(location));
+    }
 
     return cur.sub(cur.begin() + size).data();
 }
 
 hilti::rt::Bytes detail::expectBytesLiteral(
     hilti::rt::ValueReference<hilti::rt::Stream>& data, const hilti::rt::stream::View& cur, hilti::rt::Bytes literal,
```

Comment:
Fly-by as I had mentioned this on the original issue: If IIUC this results in a string allocation for something that in generated code is just a constant. If the literal is large enough and short-string optimization not in effect, this results in a malloc() and copy (might be visible generating a parser with a longer and longer literal in use).

Reply:
I saw the note but I don't think this has changed (didn't check but I thought we were creating an instance in the original code too); and it didn't seem to be the culprit, so didn't look further into it. I think this would fall into the "optimization" category: the general machinery expects a `Bytes` at runtime because that's how these values are represented, but yes, we should be able to optimize this specific case. If you think it'd be worth it, yes, file a new ticket and I'll look into it.